src/main.ts

```diff
@@ -51,128 +52,105 @@ class RubyDebugSession extends DebugSession {
-				that.sendEvent(new OutputEvent(data.toString() + '', 'stderr'));
+				this.sendEvent(new OutputEvent(data.toString() + '\n', 'console'));
 			});
 		}
-
-		//this.sendResponse(response);
-
-		if (args.stopOnEntry) {
-			this.sendResponse(response);
-
-			// we stop on the first line
-			this.sendEvent(new StoppedEvent('entry', RubyDebugSession.THREAD_ID));
-		}
 	}
 
 	// Executed after all breakpints have been set by VS Code
-	protected configurationDoneRequest(response: DebugProtocol.ConfigurationDoneResponse, args:
+	protected configurationDoneRequest(response: DebugProtocol.ConfigurationDoneRequest, args:
 	DebugProtocol.ConfigurationDoneArguments): void {
-		var command = ['start'];
-		this.rubyProcess.Run(command.join(' ') + '\n');
+		this.rubyProcess.Run('start');
+	}
+	protected setExceptionBreakPointsRequest(response: DebugProtocol.SetExceptionBreakpointsResponse, args: DebugProtocol.SetExceptionBreakpointsArguments): void {
+		if (args.filters.indexOf('all') >=0){
+			//Exception is the root of all (Ruby) exceptions - this is the best we can do
+			//If someone makes their own exception class and doesn't inherit from
+			//Exception, then they really didn't expect things to work properly
+			//anyway.
+
+			//We don't do anything with the return from this, but we
+			//have to add an expectation for the output.
+			this.rubyProcess.Enqueue('catch Exception').then(()=>1);
+		}
+		else {
+			this.rubyProcess.Run('catch off');
+		}
 		this.sendResponse(response);
 	}
-
 	protected setBreakPointsRequest(response: DebugProtocol.SetBreakpointsResponse, args: DebugProtocol.SetBreakpointsArguments): void {
-
-		var that = this;
 		var path = args.source.path;
-		var clientLines = args.lines;
-
-		// read file contents into array for direct access
-		var lines = readFileSync(path).toString().split('\n');
 
-		var linesToAdd = args.breakpoints.map(b => b.line);
-		var registeredBks = this._breakPoints.get(path);
-		var linesToRemove: number[] = [];
-		var linesToUpdate: number[] = linesToAdd;
-
-		if (registeredBks) {
-			linesToRemove = registeredBks.map(b => b.line).filter(oldLine => linesToAdd.indexOf(oldLine) === -1);
-		    linesToUpdate = registeredBks.map(b => b.line).filter(oldLine => linesToAdd.indexOf(oldLine) >= 0);
+		//to ensure that breakpoints with altered conditions are set, it is
+		//best to remove all and add them back.
+		if (this._breakPoints.get(path)) {
+			this.rubyProcess.Run('delete');
 		}
-
-		var linesToRemovePromise = linesToRemove.map(line => {
-			var bk = registeredBks.filter(b=> b.line === line)[0];
-			return that.rubyProcess.Enqueue('delete ' + bk.id + '\n');
+		var linesToUpdatePromise = args.breakpoints.map(b=>{
+			let conditionString =  b.condition ? ' if '+b.condition : '';
+			return this.rubyProcess.Enqueue('break '+ path+':'+b.line+conditionString);
 		});
+		Promise.all(linesToUpdatePromise).then( values => {
+			let breakpoints = values.map(attributes => {
```

Comment:
Still adding the breakpoint here, even if a condition is set. Best to stipulate the conditions aren't honoured in docs.
Have left the code for it in, but commented out, for when it does work.
Also, renamed the promises to `breakpointAddPromises`